src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -412,26 +427,28 @@ DistType HNSWIndex<DataType, DistType>::processCandidate(
         idType *candidate_pos = node_links + j;
         idType candidate_id = *candidate_pos;
 
+        if (this->visited_nodes_handler->getNodeTag(candidate_id) == visited_tag)
+            continue;
+
         // Pre-fetch the next candidate data into memory cache, to improve performance.
         idType *next_candidate_pos = node_links + j + 1;
         __builtin_prefetch(visited_nodes_handler->getElementsTags() + *next_candidate_pos);
         __builtin_prefetch(getDataByInternalId(*next_candidate_pos));
-        if (this->visited_nodes_handler->getNodeTag(candidate_id) == visited_tag)
-            continue;
+
         this->visited_nodes_handler->tagNode(candidate_id, visited_tag);
         char *currObj1 = (getDataByInternalId(candidate_id));
 
         DistType dist1 = this->dist_func(data_point, currObj1, this->dim);
-        if (top_candidates.size() < ef || lowerBound > dist1) {
+        if (lowerBound > dist1 || top_candidates.size() < ef) {
             candidate_set.emplace(-dist1, candidate_id);
 
-            top_candidates.emplace(dist1, candidate_id);
+            emplaceToHeap(top_candidates, dist1, candidate_id);
 
             if (top_candidates.size() > ef)
                 top_candidates.pop();
 
-            if (!top_candidates.empty())
-                lowerBound = top_candidates.top().first;
+            lowerBound = std::max(top_candidates.top().first, dist1);
```

Comment:
Because we return early on `k == 0`, and that's the only case when ef can be 0, the heap must contain at least one element from the insertion two lines above